### PR TITLE
feat(router): support deep-linking to anywhere in the app

### LINKED ISSUE #2642: feat(router): deep-linking
- [x] `router.generate` takes an array like `['/foo', { id: 3 }, 'bar', {post: 4}]` and returns a URL.
  - [ ] for now, each URL must start with `./` (child), `/` (root), or `..` (parent) – later we can implement a sane default behavior for segments with no prefix
- [x] `router-link` directive uses router.generate API
- [ ] Exception thrown when route not found (see #2323)
- [ ] Ability to link to route params for the current route without affecting children: `[router-link] = "[{ order: 'asc' }]"`

### PR DESCRIPTION
**Pull request by btford**
Closes #2642

[Review on Reviewable](https://reviewable.io/reviews/angular/angular/2772)

### CODE REVIEW

**Comment by btford** [comment]
Still working on this, but early feedback welcome

**Comment by IgorMinar** [comment]
Review status: 0 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[modules/angular2/src/router/path_recognizer.ts, line 159 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2772#-Jt41Zx1p2KR7H6kpVQZ)** ([raw file](https://github.com/angular/angular/blob/f3c532e1a673cd1f83cf07861b9e7d3286da4bca/modules/angular2/src/router/path_recognizer.ts#L159)):</sup>
can you type the handler symbol and then use handler.component and handler.type instead of reflective calls? or is there some other reason (other than lack of type info) for doing it this way?

---

<sup>**[modules/angular2/src/router/path_recognizer.ts, line 187 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2772#-Jt4DEHG3bPv5NjOkwLY)** ([raw file](https://github.com/angular/angular/blob/f3c532e1a673cd1f83cf07861b9e7d3286da4bca/modules/angular2/src/router/path_recognizer.ts#L187)):</sup>
trailing slash? I though we said that we wouldn't do trailing slashes

---

<sup>**[modules/angular2/src/router/route_registry.ts, line 153 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2772#-Jt4E68eAgrqnYOQJTn_)** ([raw file](https://github.com/angular/angular/blob/f3c532e1a673cd1f83cf07861b9e7d3286da4bca/modules/angular2/src/router/route_registry.ts#L153)):</sup>
linkDsl => linkParams?

---

<sup>**[modules/angular2/src/router/route_registry.ts, line 172 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2772#-Jt4EkCMmPrlr6Tegsq3)** ([raw file](https://github.com/angular/angular/blob/f3c532e1a673cd1f83cf07861b9e7d3286da4bca/modules/angular2/src/router/route_registry.ts#L172)):</sup>
what do you need to do with params here? I thought that it was the componentRecognizer.generate that took care of params

---

<sup>**[modules/angular2/src/router/router_link.ts, line 41 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2772#-Jt44CdM2748siHH7MIx)** ([raw file](https://github.com/angular/angular/blob/f3c532e1a673cd1f83cf07861b9e7d3286da4bca/modules/angular2/src/router/router_link.ts#L41)):</sup>
isn't this linkParams rather than route?

---

<sup>**[modules/angular2/test/router/outlet_spec.ts, line 197 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2772#-Jt43wVwUyOjNnK2SvMG)** ([raw file](https://github.com/angular/angular/blob/27ad984626a9eadf03f309cc72db707af23cc378/modules/angular2/test/router/outlet_spec.ts#L197)):</sup>
:-)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2772)

File: modules/angular2/src/router/router_link.ts [hunk 1]
```diff
@@ -32,13 +32,13 @@ import {Renderer} from 'angular2/src/render/api';
...
   selector: '[router-link]',
-  properties: ['route: routerLink', 'params: routerParams'],
+  properties: ['route: routerLink'],
   lifecycle: [onAllChangesDone],
   host: {'(^click)': 'onClick()'}
 })
 export class RouterLink {
-  private _route: string;
-  private _params: StringMap<string, string> = StringMapWrapper.create();
+  private _domEl;
```
Comment (IgorMinar) [inline modules/angular2/src/router/router_link.ts]: travis is sad because this variable is unused.

**Comment by btford** [comment]
Review status: 0 of 15 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[modules/angular2/src/router/path_recognizer.ts, line 187 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2772#-Jt4DEHG3bPv5NjOkwLY)** ([raw file](https://github.com/angular/angular/blob/f3c532e1a673cd1f83cf07861b9e7d3286da4bca/modules/angular2/src/router/path_recognizer.ts#L187)):</sup>
This is joining on `/`, not joining with `/` at the end.

---

<sup>**[modules/angular2/src/router/route_registry.ts, line 172 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2772#-Jt4EkCMmPrlr6Tegsq3)** ([raw file](https://github.com/angular/angular/blob/f3c532e1a673cd1f83cf07861b9e7d3286da4bca/modules/angular2/src/router/route_registry.ts#L172)):</sup>
This TODO is totally out of date.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2772)

**Comment by btford** [comment]
<sup>**[modules/angular2/test/router/outlet_spec.ts, line 197 \[r1\]](https://reviewable.io:443/reviews/angular/angular/2772#-Jt43wVwUyOjNnK2SvMG)** ([raw file](https://github.com/angular/angular/blob/27ad984626a9eadf03f309cc72db707af23cc378/modules/angular2/test/router/outlet_spec.ts#L197)):</sup>
?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2772)

**Pull request merged**